python/pyspark/sql/connect/types.py

```diff
@@ -342,20 +343,325 @@ def from_arrow_schema(arrow_schema: "pa.Schema") -> StructType:
+            if data_type_name == "array":
+                return self._parse_array_type()
+            elif data_type_name == "map":
+                return self._parse_map_type()
+            elif data_type_name == "struct":
+                return self._parse_struct_type()
+            elif data_type_name == "interval":
+                return self._parse_interval_type()
+            else:
+                return self._parse_primitive_types(data_type_name)
+
+        raise ParseException(
+            error_class="PARSE_SYNTAX_ERROR",
+            message_parameters={"error": f"'{type_str}'", "hint": ""},
+        )
+
+    def _parse_array_type(self) -> ArrayType:
+        type_str = self._type_str[self._pos :]
+        if len(type_str) > 0 and type_str[0] == "<":
+            self._pos = self._pos + 1
+            self._lstrip()
+            element_type = self._parse_data_type()
+            remaining = self._type_str[self._pos :]
+            if len(remaining) and remaining[0] == ">":
+                self._pos = self._pos + 1
+                self._lstrip()
+                return ArrayType(element_type)
+        raise ParseException(error_class="INCOMPLETE_TYPE_DEFINITION.ARRAY", message_parameters={})
+
+    def _parse_map_type(self) -> MapType:
+        type_str = self._type_str[self._pos :]
+        if len(type_str) > 0 and type_str[0] == "<":
+            self._pos = self._pos + 1
+            self._lstrip()
+            key_type = self._parse_data_type()
+            remaining = self._type_str[self._pos :]
+            if len(remaining) > 0 and remaining[0] == ",":
+                self._pos = self._pos + 1
+                self._lstrip()
+                value_type = self._parse_data_type()
+                remaining = self._type_str[self._pos :]
+                if len(remaining) > 0 and remaining[0] == ">":
+                    self._pos = self._pos + 1
+                    self._lstrip()
+                    return MapType(key_type, value_type)
+        raise ParseException(error_class="INCOMPLETE_TYPE_DEFINITION.MAP", message_parameters={})
+
+    def _parse_struct_type(self) -> StructType:
+        type_str = self._type_str[self._pos :]
+        if len(type_str) > 0 and type_str[0] == "<":
+            self._pos = self._pos + 1
+            self._lstrip()
+            fields = self._parse_struct_fields()
+            remaining = self._type_str[self._pos :]
+            if len(remaining) > 0 and remaining[0] == ">":
+                self._pos = self._pos + 1
+                self._lstrip()
+                return StructType(fields)
+        raise ParseException(error_class="INCOMPLETE_TYPE_DEFINITION.STRUCT", message_parameters={})
+
+    def _parse_struct_fields(self, sep_with_colon: bool = True) -> List[StructField]:
+        type_str = self._type_str[self._pos :]
+        m = self.REGEXP_IDENTIFIER.match(type_str)
+        if m:
+            field_name = m.group(0).lower().strip("`").replace("``", "`")
```

Comment:
I still doubt if this is something we should manually implement ... for example, what about case sensitivity .. To do this properly, we should use antlr for Python (e.g., https://github.com/antlr/antlr4/blob/master/doc/python-target.md).

We don't have new types very often but still we add. For example, in the last couple of years, we added ANSI types, date, year, month interval, timestamp without timzone, etc. We also expose the char and varchar type that was hidden before.

Reply:
@HyukjinKwon real engines write their own parsers ;)... In this case you could add a separate tokenization step to make the actual parsing a bit easier to read, and easier to maintain, and it can take care of some case sensitivity concerns. I think adding antlr for this is overkill.

As for the worry of missing stuff. I guess having some language agnostic specification test would not be a bad thing to add.

Reply:
Now that it supports `NOT NULL` or `COMMENT`, actually this still needs to check a config:

https://github.com/apache/spark/blob/f9c8a246ecc34ef4bb93c319c7c9b6ff732c962e/python/pyspark/sql/connect/types.py#L574-L575

In Scala, it checks `spark.sql.timestampType` whether is should be `TimestampType` or `TimestampNTZType`:

https://github.com/apache/spark/blob/f9c8a246ecc34ef4bb93c319c7c9b6ff732c962e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L2878
https://github.com/apache/spark/blob/f9c8a246ecc34ef4bb93c319c7c9b6ff732c962e/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L4858-L4865